csharp_dotnetcore/13.Basic-Bot-Template/Startup.cs

```diff
@@ -0,0 +1,217 @@
+
+                // The Memory Storage used here is for local bot debugging only. When the bot
+                // is restarted, everything stored in memory will be gone.
+                IStorage dataStore = new MemoryStorage();
+
+                // Azure Blob storage.
+                // To replace with Azure Blob Storage, add the Microsoft.Bot.Builder.Azure Nuget package to your
+                // solution. The package is found at:
+                //      https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/
+                // IStorage dataStore = new Microsoft.Bot.Builder.Azure.AzureBlobStorage("AzureBlobConnectionString", "containerName");
+
+                // Create Conversation State object.
+                // The Conversation State object is where we persist anything at the conversation-scope.
+                var conversationState = new ConversationState(dataStore);
+                options.State.Add(conversationState);
+
+                // Create User State object.
+                // The User State object is where we persist anything at the user-scope.
+                var userState = new UserState(dataStore);
+                options.State.Add(userState);
+            });
+
+            // Create and register state accesssors.
+            // Acessors created here are passed into the IBot-derived class on every turn.
+            services.AddSingleton<BasicBotAccessors>(sp =>
+            {
+                var options = sp.GetRequiredService<IOptions<BotFrameworkOptions>>().Value;
+                if (options == null)
+                {
+                    throw new InvalidOperationException("BotFrameworkOptions must be configured prior to setting up the state accessors.");
+                }
+
+                var conversationState = options.State.OfType<ConversationState>().FirstOrDefault();
+                if (conversationState == null)
+                {
+                    throw new InvalidOperationException("ConversationState must be defined and added before adding conversation-scoped state accessors.");
+                }
+
+                var userState = options.State.OfType<UserState>().FirstOrDefault();
+                if (userState == null)
+                {
+                    throw new InvalidOperationException("UserState must be defined and added before adding usre-scoped state accessors.");
+                }
+
+                // Create the custom state accessor.
+                // State accessors enable other components to read and write individual properties of state.
+                var accessors = new BasicBotAccessors(userState, conversationState)
+                {
+                    GreetingStateProperty = userState.CreateProperty<GreetingState>(BasicBotAccessors.GreetingStateName),
+                    DialogStateProperty = conversationState.CreateProperty<DialogState>(BasicBotAccessors.DialogStateName),
+                };
+
+                return accessors;
+            });
```

Comment:
This feels like an anti pattern.  What is the case for having to push your bot logic into startup.cs?

Reply:
I could go either way on this - regardless, we're going to have to pass state.  If we want to recreate accessors on every turn, don't really care - not expensive.